### ISSUE TITLE
Generated code fails validation: The Object type (OpTypePointer) in OpCompositeInsert does not match the type that results from indexing into the Composite (OpTypeRuntimeArray).

### ISSUE THREAD

**Issue by dneto0** [google/clspv#14]
Code generated from Convolution Layer case in https://github.com/google/clspv/issues/11  fails validation.

```
info: 172: Duplicate non-aggregate type declarations are not allowed. Opcode: 21 error: 1251: The Object type (OpTypePointer) in OpCompositeInsert does not match the type that results from indexing into the Composite (OpTypeRuntimeArray).
```

This might actually be two distinct issues.  I've seen the duplicate non-aggregate type decl issue before, and will file that separately.  This issue will be for the type mismatch.

**Comment by dneto0** [comment]
I'm claiming this.

**Comment by dneto0** [comment]
The problem is in translating LLVM's insertvalue.

There's a pattern of code where the first insertion is into an undef struct value, and then later inserts to build up a full struct value.   The problem is there is an invalid ID generated for the undef value of that struct type.  That should have been generated earlier in the flow, but instead we get the ID for the struct *type* itself, and there is no corresponding undef value for that struct.

I need to see why.

**Comment by dneto0** [comment]
This is more subtle than I expected.  It looks like most kinds of undef of aggregate type are not being generated.  By a side effect of how IDs for constants are tracked (in the ValueMap, but prematurely), we end up generating invalid IDs for those undef aggregates.
(The exception looks like <4 x i8> which is handled specially).

So the fix looks like two layers:
- properly generate the undef aggregate values, and
- be defensive in recording constants into the constants list: use a separate UniqueVector rather than overloading the ValueMap.  This ought to let us see failures more easily.

**Comment by dneto0** [comment]
It's less sophisticated than that.

In GenerateSPIRVConstants, there's a loop that iterates over all the constants, generating instructions for each.
There's an early "return" in the <4 x i8> case, so we end up never generating instructions for the rest of the constants.  But the earlier pass to enumerate all the constants pre-initialized the ValueMap with an index, but that index is non-sensical now.

**Comment by dneto0** [comment]
Partial fix is in https://github.com/google/clspv/commit/19a1bada4476858cec8823225d9ac5ee1b71f5d3 
That fixes the compiler so it generates the OpUndef for the struct value.

Still remaining from the case in #11 is that the Object value on the OpCompositeInsert is of incorrect type.

**Comment by dneto0** [comment]
Attaching the convolution_layer.cl example from #11, preprocessed to include the header, and with -DDATA_TYPE=float
[dut.txt](https://github.com/google/clspv/files/1252891/dut.txt)

**Comment by dneto0** [comment]
Looking at that dut.txt, it's trying to do something we don't support:

```
typedef struct Image
{
    __global uchar *ptr;
    int offset_first_element_in_bytes;
    int stride_x;
    int stride_y;
} Image;

...
# 145 "helpers.h"
Image inline update_image_workitem_ptr(__global uchar *ptr, uint offset_first_element_in_bytes, uint stride_x, uint step_x, uint stride_y, uint step_y)
{
    Image img =
    {
        .ptr = ptr,
        .offset_first_element_in_bytes = offset_first_element_in_bytes,
        .stride_x = stride_x,
        .stride_y = stride_y
    };
    img.ptr += img.offset_first_element_in_bytes + get_global_id(0) * step_x + get_global_id(1) * step_y;
    return img;
}
```
So it's returning a struct that embeds a pointer-to-char.  The compiler is not set up to understand this, since it doesn't have general-purpose pointer-to-char.

The compiler has substituted an OpTypeRuntimeArray uint  for that pointer-to-char, and then the insertion fails to sin

```
%uint_0 = OpTypeInt 32 0
%_ptr_StorageBuffer_uint = OpTypePointer StorageBuffer %uint  
...
%_runtimearr_uint = OpTypeRuntimeArray %uint
...
 %_struct_27 = OpTypeStruct %_runtimearr_uint %uint_0 %uint_0 %uint_0    
... 

%137 = OpCompositeInsert %_struct_27 %136 %44 0

%136 = OpPtrAccessChain %_ptr_StorageBuffer_uint %121 %135 
...

Basically, it's a very confused compiler. 

But lowering the priority on this because we don't support full pointer-to-char

**Comment by rjodinchr** [comment]
Gemini:
> Structs containing pointer elements (such as the convolution layer example from #11) compile cleanly and pass SPIR-V validation without invalid OpCompositeInsert instructions.

Please reopen if this is still relevant
